keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1197 @@
+natural to users. The reason for the currently proposed strategy is to maintain
+the Job property, that it is terminated as soon as it is doomed to be failed.
+
+### Support backoffLimitPerIndex when restartPolicy=OnFailure
+
+We've considered supporting the backoffLimitPerIndex when pod's `restartPolicy=OnFailure`.
+However, in this mode it is Kubelet's responsibility to restart the pod, while
+the maximal number of restarts would be enforced by the Job controller, which
+is racy in its nature. For example, in-between the checks by the Job controller,
+Kubelet execute more restarts than the specified `.spec.backoffLimit`.
+
+We believe that this feature can be supported well by using the pod-level API,
+started in this KEP:
+[Add a new field maxRestartTimes to podSpec when running into RestartPolicyOnFailure](https://github.com/kubernetes/enhancements/issues/3322).
+
+Once the pod-level API is done, it could be considered to support `.spec.runPolicy.backoffLimitPerIndex`
+when`restartPolicy=OnFailure` in pod's spec. In this case we could set the pod-level
+`maxRestartTimes` field based on the Job-level `.spec.backoffLimit`, leaving the
+responsibility of enforcing the limit to the Kubelet.
+
+### Set backoffLimitPerIndex at the top level
+
+We could set the backoffLimitPerIndex at top-level. For example:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 4
+  backoffLimitPerIndex: true
+  ...
+```
+
+This would group backoffLimit and backoffLimitPerIndex close together. However,
+the backoffLimitPerIndex is also conceptually close to maxFailedIndexes.
+We will consider this option based on the community feedback.
+
+### Enforce backoffLimit and backoffLimitPerIndex
+
+We've also considered to support both `backoffLimit` and integer
+`backoffLimitPerIndex`. In that case the Job would be marked failed if the
+global counter of failures `backoffLimit` is reached, as well as if the
+index-scoped limit is hit.
+
+However, we don't see a natural use case for this, so it seems an unnecessary
+complication. Also, the default value for `backoffLimit` of 6 might often be
+not enough when `backoffLimitPerIndex` is used, leading to potential mistakes.
+
+### Use int field
+
+We've considered to use a int `backoffLimitPerIndex` field. Here is an example:
+
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 1
+  runPolicy:
+    backoffLimitPerIndex: 4
+  ...
+```
+
+However, it is not clear what to do with the regular `backoffLimit` then.
+Here are a couple of options, but none is tempting:
+1. respect both backoffLimit and backoffLimitPerIndex, but we
+   prefer to avoid respecting both backoff limits to avoid confusion
+   (see [Enforce backoffLimit and backoffLimitPerIndex](#enforce-backofflimit-and-backofflimitperindex)).
+2. do not set the `.spec.backoffLimit` by defaulting, then Job
```

Comment:
I don't feel strongly, but I like backoffLimitPerIndex as number. API seems more local. I was thinking of variant 2 i.e. enforce only one is specified and disable backoffLimit semantics if backoffLimitPerIndex is specified. The argument with downgrade seems tricky even for the currently proposed option -- downgrade would significantly change the semantics, so I'm not sure it counts as non-breaking.

Reply:
Yes, the semantics is very different, but it would not fail. I think the situation is similar with the introduction of Indexed Jobs: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2214-indexed-job#upgrade--downgrade-strategy. In that case also the semantics is very different, and on downgrade the regular Jobs would be used (so some applications would't get the expected index env. variable). 

if we allow for not setting backoffLimit, then then the controller itself will break on downgrade.